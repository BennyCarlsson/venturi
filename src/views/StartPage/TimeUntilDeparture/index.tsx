import Paragraph from "components/Paragraph";
import Colors from "tokens/colors";
import { FontWeight } from "types";

type TimeUntilDepartureProps = {
  date?: string;
  time?: string;
  rtime?: string;
};

const TimeUntilDeparture = ({ time, date }: TimeUntilDepartureProps) => {
  const getDays = (ms: number) => {
    const days = Math.floor(ms / 86400000);
    if (days <= 0) return "";
    if (days === 1) return `${days} day `;
    return `${days} days `;
  };

  const getHours = (ms: number) => {
    const hours = Math.floor((ms % 86400000) / 3600000);
    if (hours <= 0) return "";
    if (hours === 1) return `${hours} hour `;
    return `${hours} hours `;
  };

  const getMinutes = (ms: number) => {
    const minutes = Math.round(((ms % 86400000) % 3600000) / 60000);
    if (minutes <= 0) return "";
    if (minutes === 1) return `${minutes} minute `;
    return `${minutes} minutes `;
  };

  const getTimeUntilDeparture = () => {
    if (!date || !time) return;
    const today = new Date();
    var arr: any[] = `${date} ${time}`.split(/[- :]/);
    const departure = new Date(arr[0], arr[1] - 1, arr[2], arr[3], arr[4]);

    const diffMs = departure.valueOf() - today.valueOf();
    const days = getDays(diffMs);
    const hours = getHours(diffMs);
    const minutes = getMinutes(diffMs);
    if (diffMs <= 60000) return "now";
    return `in ${days}${hours}${minutes}`;
  };

  return (
    <Paragraph fontWeight={FontWeight.bold} fontSize={14} color={Colors.vBlue}>
      {getTimeUntilDeparture()}
    </Paragraph>
  );
};

export default TimeUntilDeparture;
