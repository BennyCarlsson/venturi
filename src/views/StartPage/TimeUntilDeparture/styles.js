import styled from "styled-components";

export const TimeUntilDepartureWrapper = styled.div`
  padding: 16px;
`;
