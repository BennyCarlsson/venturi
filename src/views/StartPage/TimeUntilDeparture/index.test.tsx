import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import TimeUntilDeparture from '.';

describe('<TimeUntilDeparture />', () => {
  const date = '2021-01-01';
  const time = '13:05';
  const mockDate = new Date(2021, 0, 1, 13, 2, 0);
  jest.useFakeTimers('modern');
  jest.setSystemTime(mockDate);
  it('should match snapshot', () => {
    const tree = renderer.create(<TimeUntilDeparture date={date} time={time} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen).toBeTruthy();
  });

  it('should render minutes until departue', () => {
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('in 3 minutes')).toBeInTheDocument();
  });

  it('should render departure is now', () => {
    const date = '2021-01-01';
    const time = '13:10';
    const mockDate = new Date(2021, 0, 1, 13, 10, 0);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('now')).toBeInTheDocument();
  });

  it('should render departure is now even if late', () => {
    const date = '2021-01-01';
    const time = '13:10';
    const mockDate = new Date(2021, 0, 1, 13, 11, 0);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('now')).toBeInTheDocument();
  });

  it('should render departure is now seconds', () => {
    const date = '2021-01-01';
    const time = '13:10';
    const mockDate = new Date(2021, 0, 1, 13, 9, 30);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('now')).toBeInTheDocument();
  });

  it('should render everything singular', () => {
    const date = '2021-01-02';
    const time = '14:10';
    const mockDate = new Date(2021, 0, 1, 13, 8, 55);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('in 1 day 1 hour 1 minute')).toBeInTheDocument();
  });

  it('should render everything plural', () => {
    const date = '2021-01-03';
    const time = '15:20';
    const mockDate = new Date(2021, 0, 1, 13, 8, 55);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('in 2 days 2 hours 11 minutes')).toBeInTheDocument();
  });

  it('should render only hours', () => {
    const date = '2021-01-01';
    const time = '15:20';
    const mockDate = new Date(2021, 0, 1, 13, 20, 0);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('in 2 hours')).toBeInTheDocument();
  });

  it('should render only days', () => {
    const date = '2021-01-03';
    const time = '13:00';
    const mockDate = new Date(2021, 0, 1, 13, 0, 0);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
    render(<TimeUntilDeparture date={date} time={time} />);
    expect(screen.getByText('in 2 days')).toBeInTheDocument();
  });

  // rtime?
});
