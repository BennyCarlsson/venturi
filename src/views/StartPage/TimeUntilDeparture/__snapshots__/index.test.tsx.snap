// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeUntilDeparture /> should match snapshot 1`] = `
<p
  className="sc-bdfBwQ dmwOly"
  color="#719FB0"
  fontSize={14}
  fontWeight={700}
>
  in 3 minutes 
</p>
`;
